### PR TITLE
[FIX] hr_payroll miscalculation in no of work days #17249. 

### PR DESCRIPTION
**Pull request by jordanvrtanoski**
[FIX] hr_payroll miscalculation in no of work days when a works schedule with irregular time-slices is used #17249.
[TEST] adding test cases to reproduce the defect #17249

Description of the issue/feature this PR addresses:
Issue #17249. The new implementation for the calculation of the working days ignores the Work Schedule. This causes problems with installation where work schedule is not multiple of 8 working hours.
Current behavior before PR:
The number of days calculated in the WORK100 attendance line is greater than the real working days, sometimes even greater than the available calendar days in the period.
Desired behavior after PR is merged:
The new code is taking the Work Schedule in the consideration when the WORK100 days and leave days are calculated. To achieve this, there was a major restructuring of the previous method.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by tde-banana-odoo** [comment]
Hello,

Thanks for the proposal ! This is quite a complex issue to try to make payroll compute hours and days correctly in v10 :)  Glad someone tries to do it.

In current master version (pre-v11) computation should already be fixed as it relies completely on resource module. This one is in charge of computing the number of worked hours and days. That computation does not depend on an hardocded 8 hours/day anymore.

Fixing it in stable version is effectively a good idea. We cannot backport the master improvements as it was related to a small refactoring of the resource module. I looked at your patch and I wonder if we cannot lessen the diff.

If you look at get_working_intervals_of_day (https://github.com/odoo/odoo/blob/a6025daaf089d18c90a03199d1ea3a04d5b0945b/addons/resource/models/resource.py#L264 ) you see that you can give a compute_leaves parameter. Setting it to True should return the number of effectively worked hours that day taking leave requests into consideration. Maybe you could compute the ratio between the total number and effective hours for a given day and compute data['number_of_days'] based on that. I think we could improve the computation with a small diff. Does that seem feasible ? If so maybe you could give it a try.

Thanks,

Thibault.

**Comment by jordanvrtanoski** [comment]
Thanks, I will check the get_working_intervals_of_day and see how I can put some of the implementation there and try to prepare the fix to have less diff from pre-v11.

**Comment by jordanvrtanoski** [comment]
@tde-banana-odoo I have implemented the logic in a method _get_work_intervals_and_leaves_ on _hr.contract_ which returns the intervals of work time and vacation. The logic to get the leaves is implemented in separate method under name _get_resource_leaves_ which in pre_v11 can be dropped and we can use _iter_leaves_ instead.

File: addons/hr_payroll/models/hr_payroll.py [hunk 1]
```diff
@@ -97,6 +97,130 @@ def get_all_structures(self):
...
+        fields.Datetime.from_string(x.date_from), fields.Datetime.from_string(x.date_to), x.holiday_status_id.name),
+                              leave_ids)
+
+        day_from = fields.Datetime.from_string(date_from)
+        day_to = fields.Datetime.from_string(date_to)
+        nb_of_days = (day_to - day_from).days + 1
+
+        # Gather all intervals and holidays
+        for day in range(0, nb_of_days):
+            working_intervals_on_day = self.working_hours.get_working_intervals_of_day(
```
Comment (jordanvrtanoski) [inline addons/hr_payroll/models/hr_payroll.py]: We need to have all the intervals for the day so we can calculate the scheduled working hours

File: addons/hr_payroll/models/hr_payroll.py [hunk 2]
```diff
@@ -97,6 +97,130 @@ def get_all_structures(self):
...
+                start_dt=day_from + timedelta(days=day))
+            total_working_hours_of_day = sum(map(lambda x: hours_in_interval(x), working_intervals_on_day))
+
+            for interval in working_intervals_on_day:
+                intervals_removed_leaves = self.working_hours.interval_remove_leaves(interval, leave_intervals)
```
Comment (jordanvrtanoski) [inline addons/hr_payroll/models/hr_payroll.py]: We are cleaning the schedule intervals from the leaves in the interval so we will have the real working hours.

File: addons/hr_payroll/models/hr_payroll.py [hunk 3]
```diff
@@ -97,6 +97,130 @@ def get_all_structures(self):
...
+            """
+                Compute the duration in hours of the intreval
+            :param interval: Interval tuple (date from, date to)
+            :return: Total hours in the interval
+            """
+            return (interval[1] - interval[0]).total_seconds() / 3600.0
+
+        res = []
+        # Read all leaves for the period
+        leave_ids = self.get_resource_leaves(date_from, date_to)
```
Comment (jordanvrtanoski) [inline addons/hr_payroll/models/hr_payroll.py]: This and following 3 lines can be replaced in pre_v10 with the _employee_id.iter_leaves_

**Comment by tivisse** [comment]
Thanks for your work !

An internal task has been created to manage this case in the payroll module.

Have a nice day,

Yannick.

**Comment by tivisse** [comment]
Hello,

Thank you for the report but I will have to reject this, sorry.

The 10.0 is now in end of life cycle and will be unsupported in less than 1 year. We are only patching critical or security issues. The risk of regressions or introducing side effects is too high.

For your information the issue has been fixed in saas-11.2 by modifying the resource module. A field hours_per_day (by default set to 8) is introduced to correctly compute the working hours in case of leaves. See: https://github.com/odoo/odoo/commit/8386773a5b431accc484b7a0e2b27ed04cc9735b#diff-e7065edb86cab11d451cdfb0898d01ce
 
Regards,

Yannick.

**Pull request closed without merging**